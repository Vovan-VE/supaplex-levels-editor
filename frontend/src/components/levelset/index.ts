export * from "./FlushIndicator";
export * from "./LevelsetEditor";
export * from "./LevelsTabs";
export * from "./LevelsToolbar";
export * from "./fmt";
