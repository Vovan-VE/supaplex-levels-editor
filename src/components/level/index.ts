export * from "./body";
export * from "./toolbars";
