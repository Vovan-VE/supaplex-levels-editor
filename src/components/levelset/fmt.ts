export const fmtLevelNumber = (index: number, maxDigits: number) =>
  String(index + 1).padStart(maxDigits, "0");

const fmtLevelTitle = (title: string) =>
  title.replace(/^[-\s=[\](){}<>~]+|[-\s=[\](){}<>~]+$/g, "");

export const fmtLevelShort = (
  index: number,
  maxDigits: number,
  title: string,
) => `${fmtLevelNumber(index, maxDigits)}: ${fmtLevelTitle(title)}`;

export const fmtLevelFull = (index: number, maxDigits: number, title: string) =>
  `${fmtLevelNumber(index, maxDigits)}: ${title}`;
