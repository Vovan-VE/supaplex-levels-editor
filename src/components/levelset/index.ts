export * from "./fmt";
export * from "./LevelsetEditor";
export * from "./LevelsTabs";
export * from "./LevelsToolbar";
