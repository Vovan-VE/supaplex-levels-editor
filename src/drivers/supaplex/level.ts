import { ISupaplexLevel, ISupaplexSpecPortProps } from "./types";
import { LevelBody } from "./body";
import { BODY_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH, supaplexBox } from "./box";
import { FOOTER_BYTE_LENGTH, LevelFooter, TITLE_LENGTH } from "./footer";
import { isSpecPort } from "./tiles";

export const LEVEL_BYTES_LENGTH = BODY_LENGTH + FOOTER_BYTE_LENGTH;

const sliceFooter = (data?: Uint8Array) => {
  if (data) {
    if (
      process.env.NODE_ENV !== "production" &&
      data.length !== LEVEL_BYTES_LENGTH
    ) {
      throw new Error(
        `Invalid buffer length ${data.length}, expected ${LEVEL_BYTES_LENGTH}`,
      );
    }
    return data.slice(BODY_LENGTH);
  }
};

export class SupaplexLevel extends LevelFooter implements ISupaplexLevel {
  static fromArrayBuffer(buffer: ArrayBufferLike, byteOffset?: number) {
    return new SupaplexLevel(
      new Uint8Array(buffer, byteOffset, LEVEL_BYTES_LENGTH),
    );
  }

  readonly #body: LevelBody;

  constructor(data?: Uint8Array) {
    super(LEVEL_WIDTH, sliceFooter(data));

    this.#body = new LevelBody(supaplexBox, data?.slice(0, BODY_LENGTH));
  }

  copy() {
    return new SupaplexLevel(this.raw);
  }

  get length() {
    return BODY_LENGTH + super.length;
  }

  get raw() {
    const result = new Uint8Array(this.length);
    result.set(super.getRaw(LEVEL_WIDTH), BODY_LENGTH);
    result.set(this.#body.raw, 0);
    return result;
  }

  get width() {
    return LEVEL_WIDTH;
  }

  get height() {
    return LEVEL_HEIGHT;
  }

  get isResizable() {
    return false;
  }

  getCell(x: number, y: number) {
    return this.#body.getCell(x, y);
  }

  setCell(x: number, y: number, value: number) {
    this.#body.setCell(x, y, value, (prev) => {
      if (isSpecPort(prev)) {
        if (!isSpecPort(value)) {
          this.deleteSpecPort(x, y);
        }
      } else {
        if (isSpecPort(value)) {
          this.setSpecPort(x, y);
        }
      }
    });
  }

  get maxTitleLength() {
    return TITLE_LENGTH;
  }

  findSpecPort(x: number, y: number) {
    supaplexBox.validateCoords?.(x, y);
    return super.findSpecPort(x, y);
  }

  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps) {
    supaplexBox.validateCoords?.(x, y);
    super.setSpecPort(x, y, props);
  }

  deleteSpecPort(x: number, y: number) {
    supaplexBox.validateCoords?.(x, y);
    super.deleteSpecPort(x, y);
  }
}
