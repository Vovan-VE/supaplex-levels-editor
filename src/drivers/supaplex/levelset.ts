import { ISupaplexLevel, ISupaplexLevelset } from "./types";
import { SupaplexLevel } from "./level";
import { IBaseLevelset } from "../types";

const validateLevelsCount =
  process.env.NODE_ENV === "production"
    ? undefined
    : (count: number) => {
        if (count < 1) {
          throw new RangeError(`Invalid levels count ${count}`);
        }
      };

const validateLevelsIndex =
  process.env.NODE_ENV === "production"
    ? undefined
    : (index: number, count: number) => {
        if (index < 0 || index >= count) {
          throw new RangeError(`Invalid level index ${index}`);
        }
      };

export class Levelset<L extends ISupaplexLevel> implements IBaseLevelset<L> {
  readonly #levels: L[];

  constructor(levels: readonly L[] | Iterable<L>) {
    this.#levels = [...levels];
    validateLevelsCount?.(this.#levels.length);
  }

  get levelsCount() {
    return this.#levels.length;
  }

  get minLevelsCount() {
    return 1;
  }

  get maxLevelsCount(): number | null {
    return null;
  }

  *getLevels() {
    yield* this.#levels;
  }

  getLevel(index: number) {
    validateLevelsIndex?.(index, this.levelsCount);
    return this.#levels[index];
  }

  setLevel(index: number, level: L) {
    validateLevelsIndex?.(index, this.levelsCount);
    // TODO: dupe reference check, copy from input or remove method?
    this.#levels[index] = level;
  }

  appendLevel(level: L) {
    this.#levels.push(level);
  }

  insertLevel(index: number, level: L) {
    validateLevelsIndex?.(index, this.levelsCount);
    this.#levels.splice(index, 0, level);
  }

  removeLevel(index: number) {
    validateLevelsIndex?.(index, this.levelsCount);
    if (process.env.NODE_ENV !== "production" && this.levelsCount === 1) {
      throw new RangeError(`Cannot remove the last one level`);
    }
    this.#levels.splice(index, 1);
  }
}

const newLevels = (count: number) => {
  validateLevelsCount?.(count);
  const result: ISupaplexLevel[] = [];
  for (let i = count; i-- > 0; ) {
    result.push(new SupaplexLevel());
  }
  return result;
};

export class SupaplexLevelset
  extends Levelset<ISupaplexLevel>
  implements ISupaplexLevelset
{
  constructor(
    levels: readonly ISupaplexLevel[] | Iterable<ISupaplexLevel> | number = 111,
  ) {
    super(typeof levels === "number" ? newLevels(levels) : levels);
  }
}
