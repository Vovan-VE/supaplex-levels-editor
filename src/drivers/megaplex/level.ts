import { ILevelBody } from "../supaplex/internal";
import { LevelBody } from "../supaplex/body";
import { FOOTER_BYTE_LENGTH, TITLE_LENGTH } from "../supaplex/footer";
import { isSpecPort, TILE_SPACE } from "../supaplex/tiles";
import { ISupaplexSpecPortProps } from "../supaplex/types";
import { AnyBox } from "./box";
import { LevelFooter } from "./footer";
import { IMegaplexLevel } from "./types";

const sliceFooter = (width: number, height: number, data?: Uint8Array) => {
  if (data) {
    const bodyLength = width * height;
    const minLength = bodyLength + FOOTER_BYTE_LENGTH;
    if (process.env.NODE_ENV !== "production" && data.length < minLength) {
      throw new Error(
        `Invalid buffer length ${data.length}, expected at least ${minLength}`,
      );
    }
    return data.slice(bodyLength);
  }
};

export class MegaplexLevel extends LevelFooter implements IMegaplexLevel {
  readonly #box: AnyBox;
  #body: ILevelBody;

  constructor(width: number, height: number, data?: Uint8Array) {
    super(width, sliceFooter(width, height, data));
    this.#box = new AnyBox(width, height);
    this.#body = new LevelBody(this.#box, data?.slice(0, this.#box.length));
  }

  get length() {
    return this.#box.length + super.length;
  }

  get raw() {
    const result = new Uint8Array(this.length);
    result.set(super.getRaw(this.#box.width), this.#box.length);
    result.set(this.#body.raw, 0);
    return result;
  }

  get body() {
    return this.#body;
  }

  get width() {
    return this.#box.width;
  }

  get height() {
    return this.#box.height;
  }

  get resizable() {
    return {
      maxWidth: 0x7fff,
      maxHeight: 0x7fff,
    };
  }

  copy(): this {
    return new MegaplexLevel(
      this.#box.width,
      this.#box.height,
      this.raw,
    ) as this;
  }

  resize(width: number, height: number) {
    const origSpecPorts = [...this.getSpecPorts()];

    let src = this;
    // reducing any dimension with spec port deletion
    if (
      (width < this.#box.width && origSpecPorts.some((p) => p.x >= width)) ||
      (height < this.#box.height && origSpecPorts.some((p) => p.y >= height))
    ) {
      src = this.copy();
      // delete spec ports entry
      for (let p of origSpecPorts) {
        if (
          (width < this.#box.width && p.x >= width) ||
          (height < this.#box.height && p.y >= height)
        ) {
          src = src.setTile(p.x, p.y, TILE_SPACE);
        }
      }
    }

    const temp = new Uint8Array(width * height + src.length - src.#box.length);
    temp.set(src.getRaw(width), width * height);
    let result = new MegaplexLevel(width, height, temp) as this;

    for (let y = Math.min(height, this.#box.height); y-- > 0; ) {
      for (let x = Math.min(width, this.#box.width); x-- > 0; ) {
        result = result.setTile(x, y, src.getTile(x, y));
      }
    }

    return result;
  }

  getTile(x: number, y: number) {
    return this.#body.getTile(x, y);
  }

  setTile(x: number, y: number, value: number) {
    const prev = this.#body.getTile(x, y);
    let next = this.copy();
    next.#body = next.#body.setTile(x, y, value);
    if (isSpecPort(prev)) {
      if (!isSpecPort(value)) {
        next = next.deleteSpecPort(x, y);
      }
    } else {
      if (isSpecPort(value)) {
        next = next.setSpecPort(x, y);
      }
    }
    return next;
  }

  get maxTitleLength() {
    return TITLE_LENGTH;
  }

  findSpecPort(x: number, y: number) {
    this.#box.validateCoords?.(x, y);
    return super.findSpecPort(x, y);
  }

  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps) {
    this.#box.validateCoords?.(x, y);
    return super.setSpecPort(x, y, props);
  }

  deleteSpecPort(x: number, y: number) {
    this.#box.validateCoords?.(x, y);
    return super.deleteSpecPort(x, y);
  }
}
