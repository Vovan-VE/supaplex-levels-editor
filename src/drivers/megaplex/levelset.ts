import { Levelset } from "../supaplex/levelset";
import { fillLevelBorder } from "../supaplex/fillLevelBorder";
import { IMegaplexLevel, IMegaplexLevelset } from "./types";
import { createLevel } from "./level";

const newLevels = (count: number) => {
  const result: IMegaplexLevel[] = [];
  const level = fillLevelBorder(createLevel(60, 24));
  for (let i = count; i-- > 0; ) {
    result.push(level);
  }
  return result;
};

export const createLevelset = (
  levels: readonly IMegaplexLevel[] | Iterable<IMegaplexLevel> | number = 1,
): IMegaplexLevelset => new MegaplexLevelset(levels);

class MegaplexLevelset
  extends Levelset<IMegaplexLevel>
  implements IMegaplexLevelset
{
  constructor(
    levels: readonly IMegaplexLevel[] | Iterable<IMegaplexLevel> | number = 1,
  ) {
    super(typeof levels === "number" ? newLevels(levels) : levels);
  }

  copy(): this {
    return new MegaplexLevelset(this.getLevels()) as this;
  }

  get maxLevelsCount(): number | null {
    return 0x7fff;
  }
}
