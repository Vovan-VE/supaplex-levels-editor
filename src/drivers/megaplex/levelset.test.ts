import { MegaplexLevel } from "./level";
import { MegaplexLevelset } from "./levelset";
import { dumpLevelset } from "./helpers.dev";

describe("levelset", () => {
  describe("constructor", () => {
    it("none", () => {
      const levelset = new MegaplexLevelset();
      expect(dumpLevelset(levelset)).toMatchSnapshot();
    });

    it("number", () => {
      const levelset = new MegaplexLevelset(2);
      expect(dumpLevelset(levelset)).toMatchSnapshot();
    });

    it("levels", () => {
      const levelset = new MegaplexLevelset([
        new MegaplexLevel(5, 3),
        new MegaplexLevel(7, 9),
      ]);
      expect(dumpLevelset(levelset)).toMatchSnapshot();
    });
  });
});
