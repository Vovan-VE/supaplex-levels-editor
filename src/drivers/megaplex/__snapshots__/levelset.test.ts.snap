// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`levelset constructor levels 1`] = `
Object {
  "levels": Array [
    Object {
      "cells": Array [
        "     ",
        "     ",
        "     ",
      ],
      "demo": null,
      "height": 3,
      "infotronsNeed": "all",
      "initialFreezeZonks": false,
      "initialGravity": false,
      "isResizable": true,
      "maxTitleLength": 23,
      "specialPorts": Array [],
      "title": "                       ",
      "width": 5,
    },
    Object {
      "cells": Array [
        "       ",
        "       ",
        "       ",
        "       ",
        "       ",
        "       ",
        "       ",
        "       ",
        "       ",
      ],
      "demo": null,
      "height": 9,
      "infotronsNeed": "all",
      "initialFreezeZonks": false,
      "initialGravity": false,
      "isResizable": true,
      "maxTitleLength": 23,
      "specialPorts": Array [],
      "title": "                       ",
      "width": 7,
    },
  ],
  "levelsCount": 2,
  "maxLevelsCount": 32767,
  "minLevelsCount": 1,
}
`;

exports[`levelset constructor none 1`] = `
Object {
  "levels": Array [
    Object {
      "cells": Array [
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
      ],
      "demo": null,
      "height": 24,
      "infotronsNeed": "all",
      "initialFreezeZonks": false,
      "initialGravity": false,
      "isResizable": true,
      "maxTitleLength": 23,
      "specialPorts": Array [],
      "title": "                       ",
      "width": 60,
    },
  ],
  "levelsCount": 1,
  "maxLevelsCount": 32767,
  "minLevelsCount": 1,
}
`;

exports[`levelset constructor number 1`] = `
Object {
  "levels": Array [
    Object {
      "cells": Array [
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
      ],
      "demo": null,
      "height": 24,
      "infotronsNeed": "all",
      "initialFreezeZonks": false,
      "initialGravity": false,
      "isResizable": true,
      "maxTitleLength": 23,
      "specialPorts": Array [],
      "title": "                       ",
      "width": 60,
    },
    Object {
      "cells": Array [
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
      ],
      "demo": null,
      "height": 24,
      "infotronsNeed": "all",
      "initialFreezeZonks": false,
      "initialGravity": false,
      "isResizable": true,
      "maxTitleLength": 23,
      "specialPorts": Array [],
      "title": "                       ",
      "width": 60,
    },
  ],
  "levelsCount": 2,
  "maxLevelsCount": 32767,
  "minLevelsCount": 1,
}
`;
